crates/sui-core/src/authority_server.rs

```diff
@@ -235,8 +243,68 @@ impl ValidatorService {
         })
     }
 
+    pub(crate) fn check_execution_overload(
+        state: Arc<AuthorityState>,
+        msg: &SenderSignedData,
+    ) -> Result<(), tonic::Status> {
+        // Too many transactions are pending execution.
+        let execution_queue_len = state.transaction_manager().execution_queue_len();
+        if execution_queue_len >= MAX_EXECUTION_QUEUE_LENGTH {
+            return Err(SuiError::TooManyTransactionsPendingExecution {
+                queue_len: execution_queue_len,
+                threshold: MAX_EXECUTION_QUEUE_LENGTH,
+            }
+            .into());
+        }
+
+        for (object_id, queue_len) in state.transaction_manager().objects_queue_len(
+            msg.intent_message()
+                .value
+                .kind()
+                .input_objects()
+                .map_err(SuiError::from)?
+                .into_iter()
+                .map(|r| r.object_id())
+                .collect(),
+        ) {
+            // When this occurs, most likely transactions piled up on a shared object.
+            if queue_len >= MAX_PER_OBJECT_EXECUTION_QUEUE_LENGTH {
+                return Err(SuiError::TooManyTransactionsPendingOnObject {
+                    object_id,
+                    queue_len,
+                    threshold: MAX_PER_OBJECT_EXECUTION_QUEUE_LENGTH,
+                }
+                .into());
+            }
+        }
+        Ok(())
+    }
+
+    pub(crate) fn check_consensus_overload(
+        consensus_adapter: Arc<ConsensusAdapter>,
+    ) -> Result<(), tonic::Status> {
+        if !consensus_adapter.check_limits() {
+            return Err(tonic::Status::resource_exhausted(
+                "Reached maximum transactions pending in consensus. Consensus is overloaded."
+                    .to_string(),
```

Comment:
this should be a SuiError. It's good here, I will have another PR to fix it soon

Reply:
https://github.com/MystenLabs/sui/pull/10064